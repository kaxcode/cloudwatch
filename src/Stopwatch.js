import React, { Component } from 'react';
import { Row, Button, Card, Col } from 'react-materialize';
import { millisecondsToHuman } from './utils/humanizeTimer';

export default class Stopwatch extends Component {
  state = {
    counter: 0,
    clicked: false
  };

  timer = null;

  startStopwatch = () => {
    if (this.state.clicked === false) {
      clearInterval(this.timer);
      this.timer = setInterval(this.tick, 10);
      this.setState({ clicked: true });
    }
  };

  pauseStopwatch = () => {
    clearInterval(this.timer);
    this.setState({ clicked: false });
  };

  clearStopwatch = () => {
    clearInterval(this.timer);
    this.setState({ counter: 0 });
    this.setState({ clicked: false });
  };

  tick = () => {
    this.setState({
      counter: this.state.counter + 1
    });
  };

  render() {
    return (
      <Row>
        <Col s={12} m={3} offset="m2 l4">
          <Card className="white parent-container" textClassName="black-text">
            <h3>CloudWatch</h3>
            <h3>{millisecondsToHuman(this.state.counter)}</h3>
            <div className="parent-container">
              <Button
                waves="light"
                className="green timer-btn"
                onClick={this.startStopwatch}
              >
                Start
              </Button>
              <Button
                waves="light"
                className="red timer-btn"
                onClick={this.pauseStopwatch}
              >
                Pause
              </Button>
              <Button
                waves="light"
                className="yellow darken-3 timer-btn timer-clear"
                onClick={this.clearStopwatch}
              >
                Clear
              </Button>
            </div>
          </Card>
        </Col>
      </Row>
    );
  }
}
