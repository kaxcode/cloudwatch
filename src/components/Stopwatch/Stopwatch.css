.Stopwatch__Container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: 1fr 120px 200px;
  grid-template-areas:
    '.main.'
    '.controllerBtns.'
    '.message.';
}

.Stopwatch__Count {
  font-family: 'Inconsolata', cursive;
  grid-area: main;
  text-align: center;
  font-size: 15em;
  margin: 80px 0 50px 0;
}

@media screen and (max-width: 700px) {
  .Stopwatch__Count {
    font-size: 9em;
  }
}
