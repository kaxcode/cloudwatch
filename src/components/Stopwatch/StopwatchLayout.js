import React from 'react';
import { Row, Button, Card, Col } from 'react-materialize';
import { millisecondsToHuman } from '../../utils/humanizeTimer';
import PropTypes from 'prop-types';

const StopwatchLayout = (props) => {

  return (
    <Row>
      <Col s={12} m={3} offset="m2 l4">
        <Card className="white parent-container" textClassName="black-text">
          <h3>CloudWatch</h3>
          <h3>{millisecondsToHuman(props.counter)}</h3>
          <div className="parent-container">
            <Button
              waves="light"
              className="green timer-btn"
              onClick={props.start}
            >
              Start
            </Button>
            <Button
              waves="light"
              className="red timer-btn"
              onClick={props.pause}
            >
              Pause
            </Button>
            <Button
              waves="light"
              className="yellow darken-3 timer-btn timer-clear"
              onClick={props.clear}
            >
              Clear
            </Button>
          </div>
        </Card>
      </Col>
    </Row>
  );
};

StopwatchLayout.propTypes = {
  start: PropTypes.func.isRequired,
  pause: PropTypes.func.isRequired,
  clear: PropTypes.func.isRequired,
  counter: PropTypes.number
};


export default StopwatchLayout;
