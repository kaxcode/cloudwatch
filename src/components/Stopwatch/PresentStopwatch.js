import React from 'react';
import { millisecondsToHuman } from '../../utils/humanizeTimer';
import { func, number, bool, object } from 'prop-types';
import './Stopwatch.css';
import Nav from '../Nav/Nav.js';

class PresentStopwatch extends React.Component {
  componentDidMount() {
    this.forceUpdateInterval = setInterval(() => this.forceUpdate(), 50);
  }
  render() {
    return (
      <main className="Stopwatch">
        <Nav />
        <div className="Stopwatch__Container">
          <h3 className="Stopwatch__Count">
            {millisecondsToHuman(localStorage.getItem('counter'))}
          </h3>
        </div>
      </main>
    );
  }
}

PresentStopwatch.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  timeRemaining: number,
  startClicked: bool,
  location: object
};

export default PresentStopwatch;
