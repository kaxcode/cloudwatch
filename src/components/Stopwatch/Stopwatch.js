import React from 'react';
import { Row, Button, Card, Col } from 'react-materialize';
import { millisecondsToHuman } from '../../utils/humanizeTimer';
import { func, number } from 'prop-types';
import MessageBoard from '../MessageBoard/MessageBoard.js';
import './Stopwatch.css';
import Nav from '../Nav/Nav.js';

const Stopwatch = props => {
  return (
    <Row className="Stopwatch">
      <Col>
        <Nav />
      </Col>
      <Col s={12} m={3} offset="m2 l4">
        <Card className="Stopwatch__Card" textClassName="black-text">
          <h3 className="Stopwatch__Heading">CloudWatch</h3>
          <h3 className="Stopwatch__Count">
            {millisecondsToHuman(props.counter)}
          </h3>
          <div className="Stopwatch__ButtonsContainer">
            <Button
              waves="light"
              className="Stopwatch__StartBtn green"
              id="stopwatch-start"
              onClick={props.onStart}
            >
              Start
            </Button>
            <Button
              waves="light"
              className="Stopwatch__StopBtn red"
              id="stopwatch-pause"
              onClick={props.onPause}
            >
              Pause
            </Button>
            <Button
              waves="light"
              className="Stopwatch__ClearBtn yellow darken-3"
              id="stopwatch-clear"
              onClick={props.onClear}
            >
              Clear
            </Button>
          </div>
        </Card>
      </Col>
      <MessageBoard />
    </Row>
  );
};

Stopwatch.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  counter: number
};

Stopwatch.defaultProps = {
  counter: 0
};
export default Stopwatch;
