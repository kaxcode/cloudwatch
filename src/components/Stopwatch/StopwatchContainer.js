import React from 'react';
import Stopwatch from './Stopwatch';

export default class StopwatchContainer extends React.Component {
  state = {
    counter: 0,
    clicked: false
  };

  timer = null;

  componentDidMount = () => {
    // Uses localStorage / 0 to set state
    const localStorageRef = localStorage.getItem('counter');
    this.setState({ counter: parseInt(localStorageRef, 0) || 0 });
  };

  componentDidUpdate() {
    // Sets the locastorage
    localStorage.setItem('counter', this.state.counter);
  }

  componentWillUnmount() {
    // Sets counter to 0 if component is unmounted
    localStorage.setItem('counter', 0);
  }

  handleStart = () => {
    if (this.state.clicked === false) {
      clearInterval(this.timer);
      this.timer = setInterval(this.tick, 10);
      this.setState({ clicked: true });
    }
  };

  handlePause = () => {
    clearInterval(this.timer);
    this.setState({ clicked: false });
  };

  handleClear = () => {
    clearInterval(this.timer);
    this.setState({ counter: 0 });
    this.setState({ clicked: false });
  };

  tick = () => {
    this.setState({
      counter: this.state.counter + 1
    });
  };

  render() {
    return (
      <Stopwatch
        onStart={this.handleStart}
        onPause={this.handlePause}
        onClear={this.handleClear}
        counter={this.state.counter}
        clicked={this.state.clicked}
      />
    );
  }
}
