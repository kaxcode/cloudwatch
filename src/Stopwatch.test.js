import React from 'react';
import Stopwatch from './Stopwatch';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders correctly', () => {
  const tree = renderer.create(<Stopwatch />).toJSON();
  expect(tree).toMatchSnapshot();
});

jest.useFakeTimers();

test('Stopwatch runs after click', () => {
  const stopwatch = shallow(<Stopwatch />);
  stopwatch.instance().startStopwatch();

  expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 10);
});
