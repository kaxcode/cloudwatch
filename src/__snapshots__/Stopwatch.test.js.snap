// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stopwatch renders correctly 1`] = `
<Row>
  <Col
    m={3}
    offset="m2 l4"
    s={12}
  >
    <Card
      className="white parent-container"
      textClassName="black-text"
    >

      <div
        className="card-content black-text"
      >
        <div>
          <h3>
            CloudWatch
          </h3>
          <h3>
            00:00:00:00
          </h3>
          <div
            className="parent-container"
          >
            <button
              className="btn waves-effect waves-light green timer-btn"
              disabled={false}
              onClick={[Function]}
            >
              Start
            </button>
            <button
              className="btn waves-effect waves-light red timer-btn"
              disabled={false}
              onClick={[Function]}
            >
              Pause
            </button>
            <button
              className="btn waves-effect waves-light yellow darken-3 timer-btn timer-clear"
              disabled={false}
              onClick={[Function]}
            >
              Clear
            </button>
          </div>
        </div>
      </div>
      <div
        className="card-reveal"
      >
        <span
          className="card-title grey-text text-darken-4"
        >
          Clear
        </Button>
      </div>
    </Card>
  </Col>
</Row>
`;
